array notation to array range transformation (closes #1576)

rupertford opened the pull request:
Will also be implemented in PSyAD as part of the preprocessing.

rupertford commented:
Need to add tests, doc string example for transformation, update documentation and look at dotproduct transformation to see how the bounds are determined - as we currently just add lbound and ubound whereas the dotproduct transformation provides actual values from the array declaration where possible.

rupertford commented:
Also currently failing for PSyAD with `helmholtz_operator_kernel` (but seems to work for `w2_to_w1_projection_kernel`).

rupertford commented:
Ready for first review from @arporter or @sergisiso. This transformation returns the bounds of the array range if they can be determined. I think this is the right thing to do, rather than just returning ':'.

rupertford commented:
The PSyAD part has been dropped as the implementation does not work with the dotproduct transformation, which needs modifying.

arporter commented on src/psyclone/tests/psyir/transformations/arraynotation2arrayrange_trans_test.py:
```diff
@@ -0,0 +1,199 @@
+
+
+def test_range(fortran_reader, fortran_writer):
+    '''Test that an array slice remains unchanged.'''
```
As mentioned below, I think this really belongs in the validate test.

rupertford replied:
Added new validate test based on this test. Kept this test as is. Let me know if you think it should be removed.

arporter commented on src/psyclone/tests/psyir/transformations/arraynotation2arrayrange_trans_test.py:
```diff
@@ -0,0 +1,199 @@
+
+
+def test_variable(fortran_reader, fortran_writer):
+    '''Test that a array notation gets replaced with an array range when
```
s/a array notation/reference to an array/

rupertford replied:
Changed.

arporter commented on src/psyclone/tests/psyir/transformations/arraynotation2arrayrange_trans_test.py:
```diff
@@ -0,0 +1,199 @@
+    assert isinstance(upper_bound, BinaryOperation)
+    assert upper_bound.operator == BinaryOperation.Operator.UBOUND
+    assert isinstance(step, Literal)
+    assert step.value == "1"
```
For completeness, please add a test for an idx other than zero. As it's an internal method I won't ask for error checking on the value of `idx`.

rupertford replied:
Added test for idx==1.

arporter reviewed (changes requested):
Nicely done. I think this transformation needs a better name though and it also needs to be
added to the User Guide.
Apart from that there are just some small tweaks to do.
New test file covers new implementation file.
All tests and examples are OK with compilation.
There is a TODO mentioning #1576 that needs to be looked at:

    ./src/psyclone/psyir/frontend/fparser2.py:        # TODO #1576 this functionality is very similar to that needed in

Reference guide builds with no new warnings.

rupertford commented on src/psyclone/psyir/transformations/arraynotation2arrayrange_trans.py:
```diff
@@ -0,0 +1,164 @@
+# -----------------------------------------------------------------------------
+# Author R. W. Ford
+
+'''Module providing a transformation from PSyIR Array notation to an
```
Changed the name and updated the description.

rupertford commented on src/psyclone/psyir/transformations/arraynotation2arrayrange_trans.py:
```diff
@@ -0,0 +1,164 @@
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Author R. W. Ford
```
Oops.

rupertford commented on src/psyclone/psyir/transformations/arraynotation2arrayrange_trans.py:
```diff
@@ -0,0 +1,164 @@
+            interested in.
+
+        :returns: the loop bounds for this array index.
+        :rtype: (Literal, Literal, Literal) or \
```
tuple-ified.

rupertford commented on src/psyclone/psyir/transformations/arraynotation2arrayrange_trans.py:
```diff
@@ -0,0 +1,164 @@
+                f"'{type(node).__name__}'.")
+        if not node.symbol.is_array:
+            raise TransformationError(
+                f"The supplied node should be a Reference that references a "
```
Modified.

rupertford commented on src/psyclone/psyir/transformations/arraynotation2arrayrange_trans.py:
```diff
@@ -0,0 +1,164 @@
+
+    def apply(self, node, options=None):
+        '''Apply the ArrayNotation2ArrayRangeTrans transformation to the
+        specified node. The node must be a Reference. If the Reference
```
Updated.

rupertford commented on src/psyclone/psyir/transformations/arraynotation2arrayrange_trans.py:
```diff
@@ -0,0 +1,164 @@
+        :param node: a Reference node.
+        :type node: :py:class:`psyclone.psyir.nodes.Reference`
+        :param options: a dict with options for transformations.
+        :type options: Optional[Dict[str]]
```
Added.

rupertford commented on src/psyclone/tests/psyir/transformations/arraynotation2arrayrange_trans_test.py:
```diff
@@ -0,0 +1,199 @@
+    symbol = node.symbol
+    lower_bound, upper_bound, step = \
+        ArrayNotation2ArrayRangeTrans._get_array_bound(symbol, 0)
+    from psyclone.psyir.nodes import Literal, BinaryOperation
```
Oops. Moved to top.

rupertford commented on src/psyclone/tests/psyir/transformations/arraynotation2arrayrange_trans_test.py:
```diff
@@ -0,0 +1,199 @@
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Author R. W. Ford
```
colonoscopied

rupertford commented on src/psyclone/tests/psyir/transformations/arraynotation2arrayrange_trans_test.py:
```diff
@@ -0,0 +1,199 @@
+    lower_bound, upper_bound, step = \
+        ArrayNotation2ArrayRangeTrans._get_array_bound(symbol, 0)
+    assert isinstance(lower_bound, BinaryOperation)
+    assert lower_bound.operator == BinaryOperation.Operator.LBOUND
```
I think I've done that.

rupertford commented on src/psyclone/tests/psyir/transformations/arraynotation2arrayrange_trans_test.py:
```diff
@@ -0,0 +1,199 @@
+            in str(info.value))
+    with pytest.raises(TransformationError) as info:
+        trans.validate(Reference(DataSymbol("x", REAL_TYPE)))
+    assert ("The supplied node should be a Reference that references a "
```
Added.

rupertford commented:
I removed the #1576 todo in `fparser2.py` as it changes the code to use loop variables rather than ":" notation which is not relevant to what we are doing here.

rupertford commented:
Ready for next review from @arporter

arporter reviewed (changes requested):
All requested changes have been made.
UG has been extended and builds fine.
Ref. guide builds with no new warnings.
Tests and examples are all OK with compilation.

I would have merged but I've just read an email 
from Simon and that made me think about what happens when we have, e.g.:

    a%b = 0

where component `b` is an array. Does this transformation still work in that case?
If not, I think probably a TODO and an new Issue is in order, rather than prolong this one.

There's also a pylint error about an un-needed import in one of the test files.

rupertford commented:
Added a test showing that an array within a structure causes the transformation to raise an exception. Also added a new issue #1858 about supporting arrays within structures and referenced this issue in the transformation.

rupertford commented:
Removed unnecessary import from test file.

rupertford commented:
Ready for next review from @arporter

arporter commented on src/psyclone/psyir/transformations/reference2arrayrange_trans.py:
```diff
@@ -133,6 +133,7 @@ def validate(self, node, options=None):
             symbol.
 
         '''
+        # TODO issue #1858. Add support for structures containing arrays.
```
Sorry to be picky but I think we should mention this limitation in the docstring of the transformation (plus the associated issue number).

rupertford replied:
Added to the docstring.

arporter commented on src/psyclone/tests/psyir/transformations/reference2arrayrange_trans_test.py:
```diff
@@ -218,6 +217,26 @@ def test_validate_range(fortran_reader):
 
 
+def test_validate_structure(fortran_reader):
+    '''Test that a StructureReference raises an exception.'''
```
Please say that this is a limitation that will be tackled by #1858.

rupertford replied:
Added.

arporter reviewed (changes requested):
Sorry to bounce it back yet again but I just want the current limitation to be better documented.
As a slight carrot, I've brought the branch to master and the tests and examples are still all fine.

rupertford commented:
I've hopefully made the required changes. Ready for another look from @arporter.
